File: kmp.java
class kmp {
	void kmpSearch(String pat,String txt){
		int m = pat.length();
		int n = txt.length();

		int lps[] = new int[m];
		int j = 0;
		computeLPSarray(pat,m,lps);
		int i = 0;
		while(i < n){
			if(pat.charAt(j) == txt.charAt(i)){
				j++;
				i++;
			}
			if(j == m){
				System.out.println("Found pattern at index" + (i-j));
				j = lps[j-1];
			}
			else if(i<n && pat.charAt(j) != txt.charAt(i)){
				if(j != 0)
					j = lps[j-1];
				else
					i = i+1;
			}
		}
	}

	void computeLPSarray(String pat,int m,int lps[]){
		int len = 0;
		int i = 1;
		lps[0] = 0;
		while(i < m){
			if(pat.charAt(i) == pat.charAt(len)){
				len++;
				lps[i] = len;
				i++;
			}
			else{
				
				if(len != 0){
					len = lps[len-1];
				}
				else{
					
					lps[i] = len;
					i++;
				}
			}
		}
	}
	public static void main(String[] args) {
		String txt = "ABABDABACDABABCABAB";
		String pat = "ABABCABAB";
		new kmp().kmpSearch(pat,txt);
	}
}	

File: SM/kmp.cpp
#include <iostream>
#include <string.h>
using namespace std;

void preProcessiong(char *pat,int *prefix,int m){
	int len = 0;
	prefix[0] = 0;

	int i = 1;
	while(i < m){
		if(pat[i] == pat[len]){
			len++;
			prefix[i] = len;
			i++;
		}
		else{
			if(len != 0){
				len = prefix[len-1];
			}
			else{
				prefix[i] = 0;
				i++;
			}
		}
	}
	/*for (int i = 0; i < m; ++i)
	{
		cout << prefix[i];
	} */
	
}

void kmp(char *txt, char *pat){
	int n = strlen(txt);
	int m = strlen(pat);
	int pres[m];
	preProcessiong(pat,pres,m);
	int i=0;
	int j=0;
	while(i < n){
		if(txt[i] == pat[j]){
			i++;
			j++;
		}
		if(j == m){
			cout << "pattern found at:" << (i-j) << "\n";
		}
		else if(i < n && pat[j] != txt[i]){
			if(j != 0){
				j = pres[j-1];
			}
			else{
				i = i+1;
			}
		}
	}
}
	
int main()
{
	char txt[] = "ABABDABACDABABCABAB";
    char pat[] = "ABABCABAB";
    kmp(txt,pat);
    return 0;
}